_source/_posts/2023-07-28-oidc_workshop.md

```diff
@@ -0,0 +1,505 @@
+
+The sample application defines the relationship between Todo items and users in the file `schema.prisma`. Prisma uses this file to create SQLite tables for you. 
+
+With OIDC support, the app will need to store information about organizations in its database, as well as tracking the relationships between those organizations and the user accounts and Todo items. 
+
+## Add Organizations
+
+To support OpenID Connect, the app needs to know about more than just users and Todo items. It needs a new model, called the `org`, which represents an organization that users can belong to. 
+
+The `org`'s ID will be an auto-increment field for demonstration purposes, although using sequential identifiers in production could leak information about how many customers you have.
+
+The `org` is identified by its email domain, which is a `string` field. 
+
+Since your app will have many customers with different identity providers, you'll store each org's OpenID configuration in the database. This will require columns for: 
+
+- Authorization endpoint, a URL where the app will redirect users from the org so they can log in
+- Token endpoint, location on the OIDC server where the app can get an ID token
+- User info endpoint, where the app can retrieve the user's profile info like name and email
+- Client ID and client secret, which are the credentials that the app can use to authenticate to the org's OIDC server
+- API Key, for future use, if you add API access to the app for users in this org
+- Issuer, an identifier representing the OIDC server for the org
+
+Here's how organization support looks in `Schema.prisma`:
+```
+model Org {
+  id        Int     @id @default(autoincrement())
+  domain    String  @unique
+  issuer                 String @default("")
+  authorization_endpoint String @default("")
+  token_endpoint         String @default("")
+  userinfo_endpoint      String @default("")
+  client_id              String @default("")
+  client_secret          String @default("")
+  apikey                 String
+  Todo     Todo[]
+  User     User[]
+}
+```
+
+## Enhance Users
+
+The sample application already manages rudimentary user accounts, but with OIDC support, users will belong to organizations. The user model in `schema.prisma` needs a relationship to an organization, which is implemented with the `org` and `orgID` fields. 
+
+Previously, the sample app treated a user's email address as their unique identifier. This can cause problems if the user's organization changes its domain name, or if the user's email address changes when they change their name. Instead, with OIDC support, a unique identifier for each user can be supplied by their identity provider. This identifier is called `subject` in the OIDC protocol, and is guaranteed to be a unique and stable identifier for that user on that server. It's possible for multiple OIDC servers to issue the same `subject` ID, but the combination of server and subject uniquely identifies a single user.
```

Comment:
```suggestion
Previously, the sample app treated a user's email address as their unique identifier. This can cause problems if the user's organization changes its domain name, or if the user's email address changes when they change their name. Instead, with OIDC support, a unique identifier for each user can be supplied by their identity provider. This identifier is called `sub` (short for "subject") in the OIDC protocol, and is guaranteed to be a unique and stable identifier for that user on that server. It's possible for multiple OIDC servers to issue the same `sub` ID, since this is not a globally unique value, so the combination of server and subject uniquely identifies a single user.
```